### DISCUSSION TITLE
New commits not shown when using `overcommit` to manage Git hooks

### DISCUSSION THREAD

**Discussion by scottjacobsen** [arxanas/git-branchless#1324] [Q&A]
### Description of the bug

When I commit a few times and then run `git sl` I see this:
┃ ◯ aba6ab9 3m test
┃ ┃
┃ ◯ b147149 2m test2
┃ ┃
┃ ● e37e0ee 6s test3

Then when I run `git prev` I see this:
┃ ◯ aba6ab9 6m test
┃ ┃
┃ ● b147149 4m test2

Here I would still expect to see `e37e0ee` as the next commit but it seems hidden

```
> git next
No more child commits to go to after traversing 0 children.
```

Then I do:
```
> git unhide e37e0ee
Unhid commit: e37e0ee test3
> git sl
```

And then I see:
┃ ◯ aba6ab9 8m test
┃ ┃
┃ ● b147149 6m test2
┃ ┃
┃ ◯ e37e0ee 4m test3

After than I can do `git next` and `git prev` as usual. So it seems like new commits are defaulting to the `hidden` state.

### Expected behavior

I expect new commits not to be hidden.

### Actual behavior

New commits appear to be hidden after switching to a different commit.

### Version of `git-branchless`

git-branchless-opts 0.8.0

### Version of `git`

git version 2.45.1

**Comment by scottjacobsen** [comment]
It appears it is a conflict with `overcommit` which clobbers the hooks installed by this project. Is it possible to "namespace" hooks so they don't collide, or is this a git limitation?

**Reply by arxanas** [reply]
I converted this issue to a discussion since it may benefit future users who may not look at the closed issues.

**Comment by arxanas** [comment]
@scottjacobsen The lack of support for multiple hooks of a given kind is a Git limitation, but sounds like `overcommit` itself is the solution for that?

git-branchless assumes that the hook is a shell script and then tries pretty hard to insert its own configuration without disturbing other configuration: https://github.com/arxanas/git-branchless/blob/ce1aa49015eacf68e1c3bfdd9dd0635bba9c5069/git-branchless-init/src/lib.rs#L151-L173

There is in principle support for using different hook management systems, if you wanted to try to add support for `overcommit`: https://github.com/arxanas/git-branchless/blob/ce1aa49015eacf68e1c3bfdd9dd0635bba9c5069/git-branchless-init/src/lib.rs#L107-L121

As an ad-hoc solution, it seems like you should be able to add the git-branchless hooks to `overcommit`: https://github.com/sds/overcommit?tab=readme-ov-file#adding-existing-git-hooks

- unfortunately, there is no formal documentation for the list of hooks that git-branchless installs, but you could find the list here: https://github.com/arxanas/git-branchless/blob/ce1aa49015eacf68e1c3bfdd9dd0635bba9c5069/git-branchless/tests/test_init.rs#L226
- I believe they're all invoked via `git-branchless hook <name> <args>`. The hook interface can be found in the Git docs, or inferred from the CLI configuration here: https://github.com/arxanas/git-branchless/blob/ce1aa49015eacf68e1c3bfdd9dd0635bba9c5069/git-branchless-opts/src/lib.rs#L181-L236

  - note that some of those are not Git hooks, but instead describe re-entrant calls of git-branchless to itself in certain circumstances

**Reply by scottjacobsen** [reply]
Thanks for all the info. Ovecommit is polite enough to move the hooks it replaces to a `hooks/old-hooks` directory. It replaces these hooks: `post-checkout  post-commit    post-merge     post-rewrite`

I installed branchless first, and then overcommit, so I'm not sure if branchless would have been able to insert itself into the `overcommit` hooks. Overcommit's files are ruby files, and not shell scripts, so I would guess not unless you made it super duper smart :).

As a temporary hack I added 
```ruby
puts "#{File.basename(__FILE__)} calling branchless"
`#{File.join(__dir__, "old-hooks", File.basename(__FILE__))} #{ARGV.join(" ")}}`
puts 'called branchless'
```

to the top of all the relevant `overcommit` hooks.

Thanks for finding the `overcommit` documentation though - I'll give that method a try - it seems cleaner!
